crates/sc-proof-of-time/src/utils.rs

```diff
@@ -2,58 +2,61 @@
 
 use sp_blockchain::{HeaderBackend, Info};
 use sp_consensus::SyncOracle;
-use sp_consensus_subspace::digests::extract_pre_digest;
-use sp_runtime::traits::{Block as BlockT, Zero};
+use sp_consensus_subspace::digests::{extract_pre_digest, PotPreDigest};
+use sp_core::H256;
+use sp_runtime::traits::{Block as BlockT, NumberFor};
 use std::sync::Arc;
-use subspace_core_primitives::{NonEmptyVec, PotProof};
-use tracing::info;
+use subspace_core_primitives::{BlockHash, SlotNumber};
+use tracing::trace;
+
+/// Info extracted from the consensus tip.
+#[derive(Debug)]
+pub(crate) struct ConsensusTipInfo<Block: BlockT<Hash = H256>> {
+    /// Block hash.
+    pub(crate) block_hash: BlockHash,
+
+    /// Block number.
+    pub(crate) block_number: NumberFor<Block>,
+
+    /// Slot number for the block.
+    pub(crate) slot_number: SlotNumber,
+
+    /// The PoT from the pre digest
+    pub(crate) pot_pre_digest: PotPreDigest,
+}
 
 /// Helper to retrieve the PoT state from latest tip.
-pub(crate) async fn get_consensus_tip_proofs<Block, Client, SO>(
+pub(crate) async fn get_consensus_tip<Block, Client, SO>(
     client: Arc<Client>,
     sync_oracle: Arc<SO>,
     chain_info_fn: Arc<dyn Fn() -> Info<Block> + Send + Sync>,
-) -> Result<NonEmptyVec<PotProof>, String>
+) -> Result<ConsensusTipInfo<Block>, String>
 where
-    Block: BlockT,
+    Block: BlockT<Hash = H256>,
     Client: HeaderBackend<Block>,
     SO: SyncOracle + Send + Sync + Clone + 'static,
 {
-    // Wait for sync to complete
     let delay = tokio::time::Duration::from_secs(1);
-    info!("get_consensus_tip_proofs(): waiting for sync to complete ...");
-    let info = loop {
-        while sync_oracle.is_major_syncing() {
-            tokio::time::sleep(delay).await;
-        }
-
-        // Get the hdr of the best block hash
-        let info = (chain_info_fn)();
-        if !info.best_number.is_zero() {
-            break info;
-        }
-        info!("get_consensus_tip_proofs(): chain_info: {info:?}, to retry ...");
+    trace!("get_consensus_tip(): waiting for sync to complete ...");
+    while sync_oracle.is_major_syncing() {
         tokio::time::sleep(delay).await;
-    };
+    }
```

Comment:
Will work as a hack, but I think you should subscribe to block import notifications instead of doing sleeps here and after each block checking if it is still syncing. Though this whole function looks suspicious to me right now.

Reply:
yes, this would change further, but seems to work well in testing so far. The block import is coming (soonish) anyways for the randomness injection fro m consensus